storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/serialization/RemoteLogSegmentMetadataTransform.java

```diff
@@ -0,0 +1,98 @@
+                .setStartOffset(segmentMetadata.startOffset())
+                .setEndOffset(segmentMetadata.endOffset())
+                .setBrokerId(segmentMetadata.brokerId())
+                .setEventTimestampMs(segmentMetadata.eventTimestampMs())
+                .setMaxTimestampMs(segmentMetadata.maxTimestampMs())
+                .setSegmentSizeInBytes(segmentMetadata.segmentSizeInBytes())
+                .setSegmentLeaderEpochs(createSegmentLeaderEpochsEntry(segmentMetadata))
+                .setRemoteLogSegmentState(segmentMetadata.state().id());
+
+        return new ApiMessageAndVersion(record, record.highestSupportedVersion());
+    }
+
+    private List<RemoteLogSegmentMetadataRecord.SegmentLeaderEpochEntry> createSegmentLeaderEpochsEntry(RemoteLogSegmentMetadata data) {
+        return data.segmentLeaderEpochs().entrySet().stream()
+                   .map(entry -> new RemoteLogSegmentMetadataRecord.SegmentLeaderEpochEntry()
+                           .setLeaderEpoch(entry.getKey())
+                           .setOffset(entry.getValue()))
+                   .collect(Collectors.toList());
+    }
+
+    private RemoteLogSegmentMetadataRecord.RemoteLogSegmentIdEntry createRemoteLogSegmentIdEntry(RemoteLogSegmentMetadata data) {
+        return new RemoteLogSegmentMetadataRecord.RemoteLogSegmentIdEntry()
+                .setTopicIdPartition(
+                        new RemoteLogSegmentMetadataRecord.TopicIdPartitionEntry()
+                                .setId(data.remoteLogSegmentId().topicIdPartition().topicId())
+                                .setName(data.remoteLogSegmentId().topicIdPartition().topicPartition().topic())
+                                .setPartition(data.remoteLogSegmentId().topicIdPartition().topicPartition().partition()))
+                .setId(data.remoteLogSegmentId().id());
+    }
+
+    @Override
+    public RemoteLogSegmentMetadata fromApiMessageAndVersion(ApiMessageAndVersion apiMessageAndVersion) {
+        RemoteLogSegmentMetadataRecord record = (RemoteLogSegmentMetadataRecord) apiMessageAndVersion.message();
+        RemoteLogSegmentId remoteLogSegmentId = buildRemoteLogSegmentId(record.remoteLogSegmentId());
+
+        Map<Integer, Long> segmentLeaderEpochs = new HashMap<>();
+        for (RemoteLogSegmentMetadataRecord.SegmentLeaderEpochEntry segmentLeaderEpoch : record.segmentLeaderEpochs()) {
+            segmentLeaderEpochs.put(segmentLeaderEpoch.leaderEpoch(), segmentLeaderEpoch.offset());
+        }
+
+        RemoteLogSegmentMetadata remoteLogSegmentMetadata =
+                new RemoteLogSegmentMetadata(remoteLogSegmentId, record.startOffset(), record.endOffset(),
+                                             record.maxTimestampMs(), record.brokerId(),
+                                             record.eventTimestampMs(), record.segmentSizeInBytes(),
+                                             segmentLeaderEpochs);
+        RemoteLogSegmentMetadataUpdate rlsmUpdate
```

Comment:
Maybe I'm missing something, but there seems to be a `RemoteLogSegmentMetadata` constructor overload that also allows you to pass the `RemoteLogSegmentState` as one of the parameters: https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogSegmentMetadata.java#L99

In that case, do you still need to apply the `RemoteLogSegmentMetadataUpdate` here?

Reply:
That constructor is kept private and wanted to control what fields can be changed on the existing instance with `RemoteLogSegmentMetadataUpdate`.

Reply:
Makes sense.